### PR TITLE
optimize raft log entry fetch

### PR DESCRIPTION
**Pull request by lishihai9017**
PTAL @zhangjinpeng1987 @siddontang @BusyJay

### CODE REVIEW

**Comment by ngaut** [comment]
Any performance enhancement ?

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -319,25 +319,37 @@ impl PeerStorage {
         let mut next_index = low;
         let mut exceeded_max_size = false;
 
-        let start_key = keys::raft_log_key(self.get_region_id(), low);
-
-        if low + 1 == high {
+        if low + 8 >= high {
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: why +8 here? any benchmark shows that multi get 8 items is faster than scan?

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -319,25 +319,37 @@ impl PeerStorage {
...
             // to fetch one empty log.
             let handle = self.engine.cf_handle(CF_RAFT).unwrap();
-            match box_try!(self.engine.get_cf(handle, &start_key)) {
-                None => return Err(RaftError::Store(StorageError::Unavailable)),
-                Some(v) => {
-                    let mut entry = Entry::new();
-                    box_try!(entry.merge_from_bytes(&v));
-                    assert_eq!(entry.get_index(), low);
-                    return Ok(vec![entry]);
+            let mut keys = Vec::with_capacity((high - low) as usize);
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer_storage.rs]: If low + 1 == high, do we need construct two vector to store keys?

**Comment by lishihai9017** [comment]
benchmark see https://github.com/pingcap/rust-rocksdb/pull/59#issuecomment-304070306

File: src/raftstore/store/peer_storage.rs (hunk 1 above)
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: please define a const val for 8 here

File: src/raftstore/store/peer_storage.rs [hunk 3]
```diff
@@ -319,25 +319,31 @@ impl PeerStorage {
...
-            match box_try!(self.engine.get_cf(handle, &start_key)) {
-                None => return Err(RaftError::Store(StorageError::Unavailable)),
-                Some(v) => {
-                    let mut entry = Entry::new();
-                    box_try!(entry.merge_from_bytes(&v));
-                    assert_eq!(entry.get_index(), low);
-                    return Ok(vec![entry]);
+            for i in low..high {
+                let key = keys::raft_log_key(self.get_region_id(), i);
+                match box_try!(self.engine.get_cf(handle, &key)) {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer_storage.rs]: Please add comment explain why no need use the same snapshot of rocksdb.

Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: I think maybe we can try tailing iterator here. 

https://github.com/facebook/rocksdb/wiki/Tailing-Iterator

@zhangjinpeng1987

File: src/raftstore/store/peer_storage.rs (hunk 1 above)
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: I think `high - low <= 8` is more readable.

**Comment by zhangjinpeng1987** [comment]
LGTM, please fix ci

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

**Comment by BusyJay** [comment]
LGTM

**Comment by siddontang** [comment]
Do we need to use bloom filter for raft cf now? @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
@siddontang raft entries must exist and in most cases in memtable, so no need use bloom filter.

**Pull request merged**